### PR TITLE
add script to generate Schemastore's schema for rockcraft.yaml

### PR DESCRIPTION
**Pull request by linostar**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

Script for auto-generation of a Schemastore's schema for rockcraft.yaml

### CODE REVIEW

**Comment by linostar** [comment]
@cjdcordeiro Ready for review.

**Comment by linostar** [comment]
snap-tests are failing due to a chisel error that seems unrelated to this PR changes:

``` /rockcraft/docs/how-to/code/create-slice
+ mkdir out
+ ./chisel/chisel cut --release ./chisel-releases --root ./out openssl_bins
2024/01/13 14:35:59 Processing ./chisel-releases release...
error: chisel.yaml: cannot parse release definition: yaml: unmarshal errors:
  line 8: field v1-public-keys not found in type setup.yamlArchive
  line 10: field v1-public-keys not found in type setup.yamlRelease```

**Comment by linostar** [comment]
GH job passed: https://github.com/linostar/rockcraft/actions/runs/7530385975/job/20496635425

**Comment by tigarmo** [comment]
@linostar the spread tests are failing because the branch on your repo is too old in relation to `main`, which has the fix. Can you rebase it?

**Comment by linostar** [comment]
@tigarmo I rebased from upstream's main but doesn't seem to solve the isuue of failing spread test.

**Comment by cjdcordeiro** [comment]
Here's the fix for this problem: https://github.com/canonical/rockcraft/pull/456

File: .github/workflows/schema.yaml
Comment (cjdcordeiro) [inline .github/workflows/schema.yaml:26]: Why `/dev/null`? It would be useful to know the mismatch when an error occurs...

File: schemas/rockcraft_schema.json
Comment (cjdcordeiro) [inline schemas/rockcraft_schema.json:2]: Where does this value come from? Sounds a bit generic...

File: .github/workflows/schema.yaml
Comment (cjdcordeiro) [inline .github/workflows/schema.yaml:26]: Just being pedantic, but this should only run when the rockcraft_schema.json file already exists, otherwise, it will throw a false positive

File: schemas/rockcraft_schema.json
Comment (cjdcordeiro) [inline schemas/rockcraft_schema.json:1]: for consistency, you might want to follow the name naming convention as snapcraft -> https://github.com/snapcore/snapcraft/tree/main/schema

So it would be: `schema/rockcraft.json`

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
Thanks! This is looking good, just some minor comments.

File: .github/scripts/schema.py
Comment (linostar) [reply .github/scripts/schema.py:31]: It's not. It was a leftover.

File: .github/workflows/schema.yaml
Comment (tigarmo) [reply .github/workflows/schema.yaml:26]: but after this PR this file will always exist right?

File: .github/workflows/schema.yaml
Comment (linostar) [reply .github/workflows/schema.yaml:29]: @cjdcordeiro  Using `find` here can be considered overkill, and would also fail its purpose if somewhere in the future we have another rockcraft.json in another dir for some reason. I think a more proper way is to fail if there is no `rockcraft.json` in the assigned folder, either by adding an `else` statement to the current command, or by simplifying the whole `if` block to something like:
```test -f schema/rockcraft.json && diff generated_schema.json schemas/rockcraft.json```

Comment (cjdcordeiro) [reply .github/workflows/schema.yaml]: y the `test` line works fine.

File: schema/rockcraft.json
Comment (cjdcordeiro) [inline schema/rockcraft.json:1]: This is missing `extensions`

**Review by cjdcordeiro — COMMENTED** [review]
LGTM but just noticed the same problem as on the schema shared in the rocks channel -> `extensions`

**Review by cjdcordeiro — APPROVED** [review]
LGTM. @tigarmo can you confirm that we shouldn't see (yet) the `extensions` field in the generated schema? I thought we would after #454 , but...

File: schema/rockcraft.json
Comment (cjdcordeiro) [reply schema/rockcraft.json:1]: Ignoring this as discussed in a separate meeting.
 - we should care about it once https://github.com/canonical/rockcraft/tree/feature/12f is merged

**Comment by linostar** [comment]
Waiting for @tigarmo approval on the other changes.

**Review by tigarmo — APPROVED** [review]
nice! Thanks a lot

**Comment by tigarmo** [comment]
> LGTM. @tigarmo can you confirm that we shouldn't see (yet) the `extensions` field in the generated schema? I thought we would after #454 , but...

The `extensions` field isn't there because it's not part of the `Project` model (it's removed from the yaml during preprocessing)

**Pull request merged**
